lib/src/components/NxFileUpload/NxMultiFileUpload/NxMultiFileUpload.tsx

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2019-present Sonatype, Inc.
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
+ */
+import React, { FormEvent, forwardRef, useRef, useContext, useEffect } from 'react';
+import { faExclamationCircle } from '@fortawesome/free-solid-svg-icons';
+import classnames from 'classnames';
+import { forEach, without } from 'ramda';
+
+import SelectedFile from '../SelectedFile';
+import NxFontAwesomeIcon from '../../NxFontAwesomeIcon/NxFontAwesomeIcon';
+import NxButton from '../../NxButton/NxButton';
+import { FormAriaContext } from '../../NxForm/context';
+import { useUniqueId } from '../../../util/idUtil';
+import { Props, propTypes, MultiSelectedFileProps } from '../types';
+
+export { Props };
+
+import '../NxFileUpload.scss';
+
+function SelectedFileWrapper({file, onDismiss: onDismissProp}: MultiSelectedFileProps) {
+  const fileRef = useRef<HTMLSpanElement>(null);
+
+  function onDismiss() {
+    onDismissProp(file);
+  }
+
+  useEffect(function() {
+    if (fileRef.current) {
+      fileRef.current.scrollIntoView({block: 'nearest'});
+    }
+  }, []);
+
+  return (
+    <SelectedFile file={file} onDismiss={onDismiss} ref={fileRef}/>
+  );
+}
+
+const NxMultiFileUpload = forwardRef<HTMLDivElement, Props>(function NxMultiFileUpload(props, ref) {
+  const {
+        className: classNameProp,
+        onChange: onChangeProp,
+        files,
+        isRequired,
+        isPristine,
+        id,
+        disabled,
+        ...attrs
+      } = props,
+      isFileSelected = !!files?.length,
+      { showValidationErrors: formShowValidationErrors } = useContext(FormAriaContext),
+      showValidationErrors = formShowValidationErrors || !isPristine,
+      showError = isRequired && showValidationErrors && !isFileSelected,
+      className = classnames('nx-multi-file-upload', classNameProp),
+      noFileMessageClassName = classnames('nx-multi-file-upload__no-file-message', {
+        'nx-multi-file-upload__no-file-message--invalid': showError
+      }),
+      inputRef = useRef<HTMLInputElement>(null),
+      inputId = useUniqueId('nx-multi-file-upload-input', id),
+      selectedFilesContainerRef = useRef<HTMLDivElement>(null),
+      validationErrorId = useUniqueId('nx-multi-file-upload-validation-error');
+
+  function onChange(evt: FormEvent<HTMLInputElement>) {
+    const dataTransferObject = new DataTransfer();
```

Comment:
I think it might warrant a comment that `DataTransfer` has the only known API by which a new `FileList` can be programmatically constructed from various `File`s

Reply:
> I think it might warrant a comment that `DataTransfer` has the only known API by which a new `FileList` can be programmatically constructed from various `File`s

Comment added here: [8dbfe33](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/8dbfe33b086bab83f311ec0e334985de01af8313)